File: BirthdayParaox/BirthdayParadox.cpp
//============================================================================
// Name        : BirthdayParadox.cpp
// Author      : Rishi Khetan
// Version     :
// Copyright   : Dota 2
// Description : Hello World in C++, Ansi-style
//============================================================================

#include <iostream>
#include <iomanip>
#include <math.h>
#include <cstdlib>
using namespace std;

int BP(int n)
{
	int birthday[23];
	int i;
	for(i=0;i<23;i++)
	{
		birthday[i]=rand()%365+1;
		cout<<"Birthday of Person "<<i+1<<" :: "<<birthday[i]<<"\n";
	}

	double p=(n*(n-1))/2;
	cout<<p<<"\n";
    double u=(364.0/365.0);
    cout<<fixed<<setprecision(4)<<u*100.0<<"\n";
    double cp=pow(u,p);
    cout<<cp*100.0<<"\n";
    double cm=1-cp;
    cout<<cm*100.0<<"\n";

    int j;
    int c=0,m=0;
    for(i=0;i<23;i++)
    {
    	for(j=i+1;j<23;j++)
    	{
    		c++;
    		if( (birthday[i]==birthday[j]) )
    		{
    			cout<<"Comparison Number :: "<<c<<"\n";
    			cout<<"Birthday ["<<i<<"] = "<<birthday[i]<<"\tBirthday ["<<j<<"] = "<<birthday[j];
    			cout<<"\n Match Number :: "<<++m<<"\n";
    		}
    	}
    }
    if(m>0)
    	return 1;
    else
    	return 0;
}

int main()
{
	cout << "Gamester" << endl; // prints Gamester
	int t=0; // number of trials
	int rm=0; // result for match
	int sm=0; // successful match
	float am; // actual match
	int i;
	cout<<"\n Enter the number of trials :: \n";
	cin>>t;
	for(i=1;i<=t;i++)
	{
		cout<<"Trial Number :: "<<i<<"\n";
		rm=BP(23);
		if(rm==1)
			sm++;
		else
			cout<<"No match found in Trial Number :: "<<i<<"\n";
		am=(float)sm/(float)i;
		cout<<"\n Actual Match % ::  "<<am*100.0<<" = ("<<sm<<"/"<<i<<") \n";
	}
	cout<<"Deviation of Actual Match from the Theoertical Probability Value :: "<<fabs(((am*100.0)-50.0477))<<"\n";
	return 0;
}

File: BirthdayParadox/BirthdayParadox.java
/* Group Name :: Juggernaut
 * Group ID :: 65
 * Problem Statement :: Birthday Paradox Problem - In a set of n random people, will two of them share the same birthday
 * Assumptions :: Non-leap years consisting of 365 days are considered. A person's birthday does not fall on 29th February.
 */
import java.io.*;
import java.util.*;
public class BirthdayParadox
{
	public static int randExp(int np) //Conducts random experiment of matching birthdays of 'np' people 
	{
		int[] birthday = new int[365]; //Array for storing birthday where indices 0-364 are for days 1-365 in a non-leap year
		   //Value 0 at Birthday[index] indicates that no person has birthday on that particular day 'index'
		   //Value 1 at Birthday[index] indicates that a person has birthday on that particular day 'index'

		int matchRes=0; //Result for matching birthday
		Random rand = new Random();
		
		for(int i = 0; i < np; i++)
		{
			int randDay = rand.nextInt((365 - 1) + 1) + 1; //Generates random day for 'i' person's birthday
			if(birthday[randDay - 1] == 1)
			{
				matchRes = 1; //Successful result for matching birthdays
				break; //Match occurs as a person's birthday already falls on day 'randDay'
			}
			else
				birthday[randDay - 1] = 1;			
		} //for ends here
		return matchRes;
	} //randExp ends here
	
	public static void main(String[] args) throws IOException
	{
		BufferedReader br = new BufferedReader( new InputStreamReader( System.in ) );
		
		int numPeople; //Number of people
		int numTrials; //Number of trials for conducting random experiment of matching birthdays of 'numPeople' people
		while(true)
		{
			System.out.println("Enter the number of people :: ");
			numPeople = Integer.parseInt( br.readLine() );
			
			if(numPeople>=2)
				break;
			else
				System.out.println("Number of people should be atleast greater than or equal to 2 !! ");
		}
		
		while(true)
		{		
			System.out.println("Enter the number of trials for conducting random experiment of matching birthdays :: ");
			numTrials = Integer.parseInt( br.readLine() );
			
			if(numTrials>0)
				break;
			else
				System.out.println("Number of trials should be greater than 0 !! ");
		}		
		System.out.println("");
		
		int numMatch = 0; //Number of trials in which successful match was found
		double probActualMatch; //Probability of successful trials in which a match was found
		
		System.out.println("List of Trial Numbers in which successful match was not found :: ");
		
		for(int i = 1; i <= numTrials; i++)
		{
			int res = randExp(numPeople); //Conduct random experiment for trial number 'i'
			if(res==1)
				numMatch += res; //Increment number of successful matches
			else
				System.out.print(i+"  ");			
		} //for ends here
		System.out.println("\n");
		
		probActualMatch = (double)numMatch/(double)numTrials;
		
		int numComp=(numPeople*(numPeople-1))/2; //Number of comparisons
		System.out.println("Number of comparisons where one person's birthday is being compared against someone else's = C("+numPeople+",2) = "+numComp);
		
		System.out.print("Probability of two people having different birthdays = 364/365 = ");
		System.out.printf("%.4f",((364.0/365.0)*100.0));
		System.out.println("%");
		
		System.out.print("Probability of all "+numComp+" pairs having different birthdays = (364/365)^"+numComp+" = ");
		double probDiff=Math.pow((364.0/365.0),numComp); //Probability of all pairs having different birthdays
		System.out.printf("%.4f",probDiff*100.0);
		System.out.println("%");
		
		System.out.print("Probability of having a match (Theoretical) = 1 - ");
		System.out.printf("%.4f",probDiff*100.0);
		System.out.print("% = ");
		System.out.printf("%.4f",(1-probDiff)*100.0);
		System.out.println("%\n");
		
		
		System.out.print("Probability of Actual Match (Experimental) = "+numMatch+"/"+numTrials+" = ");
		System.out.printf("%.2f",(probActualMatch*100.0));
		System.out.println("%\n");
		
		System.out.println("Birthday Paradox holds true and is valid and two people in a set of n random people can share the same birthday ");

	} //main ends here

} //class ends here
